### PR TITLE
progress: auto-select progress based on the environment

### PR DESCRIPTION
**Pull request by mvo5**
This commit adds automatic progress bar selection based on checking
if we run on a terminal or not. When running on a terminal we use
the nice "terminalProgressBar". If that is not set we assuem
we run in a script or CI/CD environment and select plainProgressBar.

Thanks Colin for the hint about the bad integration test.

### CODE REVIEW

**Review by achilleas-k — DISMISSED** [review]

**Comment by cgwalters** [comment]
>  Ideally we would check for a terminal but with podman we never have one.

`docker/podman run -t` ?

See also https://github.com/containers/podman-bootc/pull/61/files#diff-000666f7f8953356d352cfc7388aac26f46bb3f75ef9f413b8908a7bf8eb2729R352

which was motivated by bootc having progress too.

(And that relates to the larger overlap/duplication/desire-to-bridge between bootc progress and osbuild progress in https://github.com/containers/bootc/pull/921#issuecomment-2539688815 )

**Comment by mvo5** [comment]
> > Ideally we would check for a terminal but with podman we never have one.
> 
> `docker/podman run -t` ?
> 
> See also https://github.com/containers/podman-bootc/pull/61/files#diff-000666f7f8953356d352cfc7388aac26f46bb3f75ef9f413b8908a7bf8eb2729R352
> 
> which was motivated by bootc having progress too.
> 
> (And that relates to the larger overlap/duplication/desire-to-bridge between bootc progress and osbuild progress in [containers/bootc#921 (comment)](https://github.com/containers/bootc/pull/921#issuecomment-2539688815) )

Thanks!

Uh, that is rather embarrassing, I may indeed have just written my integration test incorrectly (and not had . I'm an idiot. Well, I guess it's good, it makes things easier.

File: bib/internal/progress/progress.go [hunk 1]
```diff
@@ -66,12 +66,29 @@ type ProgressBar interface {
 	Stop()
 }
 
+func isTerminal() bool {
+	// Note that we cannot simply use "isatty()" here because
+	// "podman run -it" will not give us a full terminal to
+	// prevent TIOCSTI - this is the next best we can do.
+	for _, env := range []string{"TERM", "SSH_CONNECTION"} {
```
[the remark this answers is not in the document]
Comment (mvo5) [reply bib/internal/progress/progress.go]: Sorry, I'm  a muppet - this is done in the right way now (hopefully)

**Comment by cgwalters** [comment]
LGTM

**Review by ondrejbudai — DISMISSED** [review]
Thanks! ❤️

**Comment by ondrejbudai** [comment]
pylint seems to be complaining in the integration tests. Can you take a look, Michael?

**Review by achilleas-k — APPROVED** [review]

**Comment by chunfuwen** [comment]
I validated it by running in a script ( assuming plainProgressBar). I noticed below picture shown as order
plainprogressbar_01
plainprogressbar_02
plainprogressbar_03

Something surprise me that there are lots of: [/] Image building step in the output, which looks a little noisy

**Comment by chunfuwen** [comment]
Moreover, I didn't see that the progress for Stage org.osbuild.container-deploy  has one 100% percentage bar(until command finish), and only saw at the beginning, there is this one [stdout] [1 / 3] Stage org.osbuild.container-deploy [--------------->________________________________] 33.33%

**Comment by benblasco** [comment]
Is this in any way configurable by the user? My understanding is that this is hard-coded, whereas users may want to see the full output when manually building for learning and troubleshooting purposes. Thoughts? I will create an issue if folks think this is a valid requirement.

**Comment by achilleas-k** [comment]
> I will create an issue if folks think this is a valid requirement.

Sure, go ahead.  I think having an option for it makes sense.

**Comment by hanthor** [comment]
Yes I would be nice to have that in the build logs on GH Actions.

**Comment by mvo5** [comment]
> I validated it by running in a script ( assuming plainProgressBar). I noticed below picture shown as order
[..]
> 
> Something surprise me that there are lots of: [/] Image building step in the output, which looks a little noisy

Something funky is going on there, the "plainProgressBar" should not contain "[/]" style spinners. So far the detection is based around if the running system is a terminal (and modelled after what bootc is doing in it's progress PR). We could tweak this further. In most CI/CD pipelines and automations (e.g. via cron/systemd-timers) the detection should be correct but maybe I'm missing something, I am definitely interested in learning more about why we got "plainProgressBar" here. The commandline would be nice, it seems from the image parts are cut(?).

I opened https://github.com/osbuild/bootc-image-builder/pull/776 that adds some more documentation, sorry that this was not included earlier and caused some confusion here. 

In a nutshell: it is possible to override via "--progress=plain" to get the old behavior and we are happy to look into tweaking the auto-detection or the defaults (i.e. ensure we do what is best for our users).
